<a href="https://colab.research.google.com/github/Mahmud394/Global-Life-Expectancy/blob/main/01_loading_and_inspecting_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab 1 — Loading and Inspecting Data

**DS331 · Introduction to Data Science**

Welcome to the first lab notebook! Every data science project starts the same way: you get a dataset, you load it into Python, and you *look at it carefully* before doing anything else.

In this notebook you will learn how to:

1. [Import the libraries we need](#1)
2. [Load a dataset (from a URL, a library, or your own file)](#2)
3. [Take a first look at the data](#3)
4. [Select rows and columns](#4)
5. [Count and summarize values](#5)
6. [Preview: missing values](#6)
7. [Try it yourself ✏️](#7)

**Dataset:** the *Palmer Penguins* dataset — body measurements of 344 penguins from three species living on three islands in Antarctica. It is small, easy to understand, and perfect for learning.

<a id="1"></a>
## 1. Importing the libraries

We always start by importing the libraries we will use. The two most important ones for data work are:

- **pandas** — for loading and manipulating tabular data (think "Excel inside Python"). Imported as `pd` by convention.
- **seaborn** — a plotting library that also ships with several practice datasets. Imported as `sns` by convention.

These short aliases (`pd`, `sns`) are universal conventions — every tutorial and Stack Overflow answer uses them, so you should too.

In [34]:
import pandas as pd
import seaborn as sns

print("pandas version:", pd.__version__)
print("seaborn version:", sns.__version__)

pandas version: 2.2.2
seaborn version: 0.13.2


<a id="2"></a>
## 2. Loading a dataset

The object that holds a table in pandas is called a **DataFrame**. There are several ways to create one — here are the three you will use most often.

### 2.1 From a CSV file on the internet

`pd.read_csv()` is *the* workhorse function. You can give it a local file path **or a URL** — it works the same way.

In [2]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv"
penguins = pd.read_csv(url)

type(penguins)

pandas.core.frame.DataFrame

### 2.2 From a library

Seaborn bundles the same dataset, so this one line is equivalent — handy for practice:

```python
penguins = sns.load_dataset("penguins")
```

### 2.3 From your own file (what you'll do for your report)

When you work on **your assigned dataset** in Google Colab, you have two options:

```python
# Option A: upload the file directly into the Colab session
from google.colab import files
uploaded = files.upload()                 # a file-picker appears
df = pd.read_csv("your_dataset.csv")

# Option B: mount your Google Drive (file survives between sessions)
from google.colab import drive
drive.mount("/content/drive")
df = pd.read_csv("/content/drive/MyDrive/your_dataset.csv")
```

> ⚠️ **Common mistake:** `pd.read_csv("dataset.csv")` fails with `FileNotFoundError` if the file is not in the current folder. In Colab, uploaded files land in `/content/` — check the 📁 file panel on the left to see what is actually there.

Useful `read_csv` options you may need for your own dataset:

| Argument | When you need it |
|---|---|
| `sep=";"` | columns are separated by `;` instead of `,` |
| `header=None` | the file has no column names in the first row |
| `names=[...]` | provide your own column names |
| `encoding="latin-1"` | you get a `UnicodeDecodeError` |
| `na_values=["?", "N/A"]` | missing values are written as something unusual |

There are also `pd.read_excel()`, `pd.read_json()` and others for non-CSV formats.

<a id="3"></a>
## 3. First look at the data

**Never skip this step.** Before any analysis, you must know: How big is the table? What are the columns? What types do they have? Do the values look sensible?

### 3.1 `head()`, `tail()`, `sample()` — peek at the rows

In [3]:
penguins.head()        # first 5 rows (pass a number for more, e.g. head(10))

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,MALE
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,FEMALE
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,FEMALE
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,FEMALE


In [4]:
penguins.tail(3)       # last 3 rows

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
341,Gentoo,Biscoe,50.4,15.7,222.0,5750.0,MALE
342,Gentoo,Biscoe,45.2,14.8,212.0,5200.0,FEMALE
343,Gentoo,Biscoe,49.9,16.1,213.0,5400.0,MALE


In [5]:
penguins.sample(5, random_state=42)   # 5 random rows — good for spotting surprises

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
194,Chinstrap,Dream,50.9,19.1,196.0,3550.0,MALE
157,Chinstrap,Dream,45.2,17.8,198.0,3950.0,FEMALE
225,Gentoo,Biscoe,46.5,13.5,210.0,4550.0,FEMALE
208,Chinstrap,Dream,45.2,16.6,191.0,3250.0,FEMALE
318,Gentoo,Biscoe,48.4,14.4,203.0,4625.0,FEMALE


Each row is one penguin. The columns mix **numeric** measurements (bill length, flipper length, body mass) and **categorical** labels (species, island, sex). Most real datasets look like this.

### 3.2 `shape`, `columns`, `dtypes` — the table's structure

Note: these are **attributes**, not methods — no parentheses `()`.

In [6]:
print("shape   :", penguins.shape)        # (rows, columns)
print("rows    :", penguins.shape[0])
print("columns :", list(penguins.columns))

shape   : (344, 7)
rows    : 344
columns : ['species', 'island', 'bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g', 'sex']


In [7]:
penguins.dtypes

,0
species,object
island,object
bill_length_mm,float64
bill_depth_mm,float64
flipper_length_mm,float64
body_mass_g,float64
sex,object


- `object` usually means **text** (strings) — here the categorical columns.
- `float64` / `int64` are **numbers**.

> ⚠️ **Common mistake:** if a column that *should* be numeric (e.g. a price) shows up as `object`, something in it is not a number — maybe a `$` sign, a comma, or a `"?"` used for missing values. We will fix exactly this kind of problem in Lab 2.

### 3.3 `info()` — structure + missing values in one view

In [8]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


Read this output carefully — it answers three questions at once:

1. **344 entries** → number of rows.
2. **Non-Null Count** → e.g. `bill_length_mm` has 342 non-null values out of 344, so **2 values are missing**. The `sex` column is missing 11.
3. **Dtype** → the type of each column.

### 3.4 `describe()` — summary statistics

In [9]:
penguins.describe()

,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
count,342.000000,342.000000,342.000000,342.000000
mean,43.921930,17.151170,200.915205,4201.754386
std,5.459584,1.974793,14.061714,801.954536
min,32.100000,13.100000,172.000000,2700.000000
25%,39.225000,15.600000,190.000000,3550.000000
50%,44.450000,17.300000,197.000000,4050.000000
75%,48.500000,18.700000,213.000000,4750.000000
max,59.600000,21.500000,231.000000,6300.000000


For every **numeric** column you get the count, mean, standard deviation, min/max, and the quartiles (25%, 50% = median, 75%).

Sanity-check these numbers against common sense: body mass between 2700 g and 6300 g is plausible for penguins. If you saw a minimum of `-999`, that would scream "missing-value code"!

By default `describe()` ignores text columns. Ask for them by excluding the numeric ones:

In [10]:
penguins.describe(exclude="number")   # count, unique values, most frequent value

,species,island,sex
count,344,344,333
unique,3,3,2
top,Adelie,Biscoe,MALE
freq,152,168,168


<a id="4"></a>
## 4. Selecting rows and columns

You constantly need to zoom in on a part of the table. There are a few core patterns — learn these well.

### 4.1 Selecting columns

In [11]:
penguins["species"].head()            # one column -> a Series (a single column object)

,species
0,Adelie
1,Adelie
2,Adelie
3,Adelie
4,Adelie


In [12]:
penguins[["species", "body_mass_g"]].head()   # note the DOUBLE brackets: a list of columns -> DataFrame

,species,body_mass_g
0,Adelie,3750.0
1,Adelie,3800.0
2,Adelie,3250.0
3,Adelie,NaN
4,Adelie,3450.0


> ⚠️ **Common mistake:** `penguins["species", "body_mass_g"]` (single brackets, two names) is a `KeyError`. To select multiple columns you pass a **list**, hence the double brackets.

### 4.2 Selecting rows by position — `iloc`

`iloc` = **i**nteger **loc**ation. It works like list indexing: `df.iloc[row_positions, column_positions]`.

In [13]:
penguins.iloc[0]          # the first row

,0
species,Adelie
island,Torgersen
bill_length_mm,39.1
bill_depth_mm,18.7
flipper_length_mm,181.0
body_mass_g,3750.0
sex,MALE


In [14]:
penguins.iloc[10:15, 0:3]   # rows 10–14, first three columns

,species,island,bill_length_mm
10,Adelie,Torgersen,37.8
11,Adelie,Torgersen,37.8
12,Adelie,Torgersen,41.1
13,Adelie,Torgersen,38.6
14,Adelie,Torgersen,34.6


### 4.3 Selecting rows by label/condition — `loc`

`loc` uses **labels** (index values and column names): `df.loc[rows, columns]`.

In [33]:
penguins.loc[0:4, ["species", "island", "sex"]]

,species,island,sex
0,Adelie,Torgersen,MALE
1,Adelie,Torgersen,FEMALE
2,Adelie,Torgersen,FEMALE
3,Adelie,Torgersen,NaN
4,Adelie,Torgersen,FEMALE


### 4.4 Boolean filtering — the most useful pattern in pandas

Write a condition on a column → get a True/False mask → use it to keep only matching rows.

In [ ]:
heavy = penguins[penguins["body_mass_g"] > 5000]
print("Penguins heavier than 5 kg:", len(heavy))
heavy.head()

In [15]:
# Combining conditions: & = and, | = or, ~ = not.
# EACH condition needs its own parentheses!
heavy_gentoo_females = penguins[
    (penguins["species"] == "Gentoo")
    & (penguins["sex"] == "Female")
    & (penguins["body_mass_g"] > 5000)
]
heavy_gentoo_females.head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex


> ⚠️ **Common mistakes** with filtering:
> - Using Python's `and` / `or` instead of `&` / `|` → `ValueError: The truth value of a Series is ambiguous`.
> - Forgetting the parentheses around each condition → confusing operator-precedence errors.
> - Using `=` (assignment) instead of `==` (comparison).

### 4.5 `query()` — the same thing, written as text

Some people find this more readable; both styles are fine.

In [32]:
penguins.query("species == 'Gentoo' and body_mass_g > 5000").head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
221,Gentoo,Biscoe,50.0,16.3,230.0,5700.0,MALE
223,Gentoo,Biscoe,50.0,15.2,218.0,5700.0,MALE
224,Gentoo,Biscoe,47.6,14.5,215.0,5400.0,MALE
227,Gentoo,Biscoe,46.7,15.3,219.0,5200.0,MALE
229,Gentoo,Biscoe,46.8,15.4,215.0,5150.0,MALE


<a id="5"></a>
## 5. Counting and summarizing

### 5.1 `value_counts()` — how often does each value appear?

This is *the* first tool for exploring a categorical column.

In [31]:
penguins["species"].value_counts()

,count
species,
Adelie,152
Gentoo,124
Chinstrap,68


In [30]:
penguins["island"].value_counts(normalize=True).round(3)   # as proportions instead of counts

,proportion
island,
Biscoe,0.488
Dream,0.360
Torgersen,0.151


In [29]:
print("unique values:", penguins["species"].unique())
print("how many     :", penguins["species"].nunique())

unique values: ['Adelie' 'Chinstrap' 'Gentoo']
how many     : 3


### 5.2 Statistics of a single column

In [28]:
col = penguins["body_mass_g"]
print("mean   :", round(col.mean(), 1))
print("median :", col.median())
print("std    :", round(col.std(), 1))
print("min/max:", col.min(), "/", col.max())

mean   : 4201.8
median : 4050.0
std    : 802.0
min/max: 2700.0 / 6300.0


### 5.3 `sort_values()` — order the table

In [27]:
penguins.sort_values("body_mass_g", ascending=False).head(5)   # the 5 heaviest penguins

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
237,Gentoo,Biscoe,49.2,15.2,221.0,6300.0,MALE
253,Gentoo,Biscoe,59.6,17.0,230.0,6050.0,MALE
337,Gentoo,Biscoe,48.8,16.2,222.0,6000.0,MALE
297,Gentoo,Biscoe,51.1,16.3,220.0,6000.0,MALE
299,Gentoo,Biscoe,45.2,16.4,223.0,5950.0,MALE


### 5.4 `groupby()` — statistics *per group*

This is one of the most powerful ideas in pandas: **split** the data into groups, **apply** a calculation to each group, and **combine** the results.

Question: *what is the average body mass of each species?*

In [26]:
penguins.groupby("species")["body_mass_g"].mean().round(1)

,body_mass_g
species,
Adelie,3700.7
Chinstrap,3733.1
Gentoo,5076.0


In [25]:
# Several statistics at once with .agg()
penguins.groupby("species")["body_mass_g"].agg(["count", "mean", "min", "max"]).round(1)

,count,mean,min,max
species,,,,
Adelie,151,3700.7,2850.0,4775.0
Chinstrap,68,3733.1,2700.0,4800.0
Gentoo,123,5076.0,3950.0,6300.0


In [24]:
# Grouping by two columns
penguins.groupby(["species", "sex"])["body_mass_g"].mean().round(1)

species    sex   
Adelie     FEMALE    3368.8
           MALE      4043.5
Chinstrap  FEMALE    3527.2
           MALE      3939.0
Gentoo     FEMALE    4679.7
           MALE      5484.8
Name: body_mass_g, dtype: float64

Already a finding for a report: *Gentoo penguins are clearly heavier than the other two species, and within every species males are heavier than females.* This is what EDA is about — turning simple summaries into statements about the data.

<a id="6"></a>
## 6. Preview: missing values

`info()` hinted that some values are missing. Here is the standard one-liner to count them per column — we will deal with them properly in **Lab 2**.

In [23]:
penguins.isna().sum()

,0
species,0
island,0
bill_length_mm,2
bill_depth_mm,2
flipper_length_mm,2
body_mass_g,2
sex,11


<a id="7"></a>
## 7. Try it yourself ✏️

Practice on a different dataset: **tips** — restaurant bills and tips. Load it and answer the questions below. (Solutions are not provided on purpose — everything you need is in this notebook.)

In [17]:
tips = sns.load_dataset("tips")
tips.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


Lab 1 — Loading and Inspecting Data

**Exercise 1.** How many rows and columns does `tips` have? Which columns are numeric and which are categorical? Are any values missing?

**Exercise 2.** Show summary statistics for the numeric columns. What is the largest total bill?

**Exercise 3.** How many smokers vs. non-smokers are in the data? (Hint: `value_counts`.)

**Exercise 4.** Select all rows where the bill is above 40 dollars *and* the party size is at least 4. How many are there?

**Exercise 5.** Using `groupby`, compute the average tip for each day of the week. Which day tips best?

In [18]:
# Exercise 1
print("Rows, Columns:", tips.shape)

print("\nNumeric Columns:")
print(tips.select_dtypes(include="number").columns.tolist())

print("\nCategorical Columns:")
print(tips.select_dtypes(exclude="number").columns.tolist())

print("\nMissing Values:")
print(tips.isnull().sum())

Rows, Columns: (244, 7)

Numeric Columns:
['total_bill', 'tip', 'size']

Categorical Columns:
['sex', 'smoker', 'day', 'time']

Missing Values:
total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64


In [19]:
# Exercise 2
print(tips.describe())

print("\nLargest Total Bill:")
print(tips["total_bill"].max())

       total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

Largest Total Bill:
50.81


In [20]:
# Exercise 3
print(tips["smoker"].value_counts())

smoker
No     151
Yes     93
Name: count, dtype: int64


In [21]:
# Exercise 4
result = tips[(tips["total_bill"] > 40) & (tips["size"] >= 4)]

print(result)
print("\nTotal Rows:", len(result))

     total_bill   tip     sex smoker   day    time  size
59        48.27  6.73    Male     No   Sat  Dinner     4
95        40.17  4.73    Male    Yes   Fri  Dinner     4
142       41.19  5.00    Male     No  Thur   Lunch     5
156       48.17  5.00    Male     No   Sun  Dinner     6
197       43.11  5.00  Female    Yes  Thur   Lunch     4
212       48.33  9.00    Male     No   Sat  Dinner     4

Total Rows: 6


In [22]:
# Exercise 5
avg_tip = tips.groupby("day")["tip"].mean()

print(avg_tip)

print("\nBest Tipping Day:")
print(avg_tip.idxmax())

day
Thur    2.771452
Fri     2.734737
Sat     2.993103
Sun     3.255132
Name: tip, dtype: float64

Best Tipping Day:
Sun


/tmp/ipykernel_2757/3408852091.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  avg_tip = tips.groupby("day")["tip"].mean()


---
**Next lab:** the data is rarely as tidy as the penguins. In **Lab 2 — Data Cleaning and Preprocessing** we work with the Titanic dataset and deal with missing values, wrong types, duplicates and outliers.